### DISCUSSION TITLE
How to improve Open API documentation in Tapir

### DISCUSSION THREAD

**Discussion by pme123** [Iltotore/iron#236] [Q&A]
Hi
How can I improve the documentation of Iron Types in Tapir Open API documentation?
Here some examples:
```
type Percentage = Int :| (GreaterEqual[0] & LessEqual[100])
type SelectedPensionType = Int :| any.In[(
  11393215,
    36496024,
    36496021,
    122483321,
    110354423,
    110354433,
    110354454,
    122483323
  )]
```
This gives me:
![image](https://github.com/Iltotore/iron/assets/3437927/46a75d12-d8b2-4648-a6c8-78c7b8fa280d)

Is there a way to do this without adding any custom description or at least adding a description in a generic way?

**Comment by Iltotore** [comment]
Hello.

I'm not familiar with Tapir. As far as I understand, you want to generate OpenAPI documentation as shown in the screenshot you linked automatically using constraints. Am I right?

If so, what do you currently have to do obtain this result?
![image](https://github.com/Iltotore/iron/assets/42907886/d28d5f73-c664-4371-b8b5-d49915281975)

**Comment by pme123** [comment]
Yes something like `Constraint: GreaterEqual[0] & LessEqual[100]`
or `Constraint: One of 11393215,  36496024,..`

Looks like i have to figure out, how Tapir is generating _Value_.

**Reply by Iltotore** [reply]
I don't know how Tapir's `Value` and Open API specs generation work but I think it should be easily doable you could just create more specialized given `Value`s

EDIT: [this issue](https://github.com/softwaremill/tapir/issues/48) might be helpful
EDIT 2: [Validator](https://www.javadoc.io/doc/com.softwaremill.sttp.tapir/tapir-core_3/latest/sttp/tapir/Validator$.html) seems to be what Tapir uses to generate OpenAPI doc
